Default SNI to hostname for SSL connections

juhlig opened the pull request:
In response to #164.

While `mysql-otp` sets the `verify` option for SSL connections so that it defaults to `verify_peer`, it does not set any default for `server_name_indication`. Since `ssl:connect/3` is used in the variant that upgrades a TCP socket to SSL, according to the documentation, this causes the SNI option to default to the IP address of the peer, which is probably not what is wanted.

This PR sets up a default for `server_name_indication` as well, defaulting it to the `host` parameter that may be given in the options to `mysql:start_link/1`. From an outside perspective, this has the benefit that the SSL connection process of `mysql-otp` behaves more like `ssl:connect/3` in the variant where it _open_ a SSL connection (vs upgrading an already connected TCP socket) and `server_name_indication` defaults to the given host name (instead of the peer IP address).

If the host given in the `mysql:start_link/1` options is itself an IP address (either in tuple or string form) or an atom (like `loopback`), `server_name_indication` defaults to `disable`. This is different from the `ssl:connect/3` behavior (as outlined above). While this is arguably the more intuitive way to go, it differs from what `ssl` does and would need extra documentation. So I'm putting this up for debate: if the host given in `mysql:start_link/1` is an IP address, should `server_name_indication` default to `disable` (and document that), or should no default be set then and `ssl` be allowed to process it the way it usually does (no additional documentation needed then)?

zuiderkwast commented on src/mysql_protocol.erl:
```diff
@@ -406,26 +408,37 @@ server_version_to_list(ServerVersion) ->
+        {error, einval} -> Host
+    end;
+get_default_sni(_) ->
+    disable.
```
Also here, I'm not sure we should use `disable` (unless we match explicitly on `{local, _}` for unix sockets perhaps, but why use TLS for unix sockets?) A `gen_tcp` host can also be an atom, which I think means non-fully qualified domain name.

What I mean is that adding `disable` potentially lowers the security, which I think we shouldn't do by default if there is any doubt. It can be done by the user.

juhlig replied:
Changed in last commit.

The atom form of the hostname, I had to dig around for that a bit, but finally found out that it can (only?) be the local hostname. I think this doesn't make much sense, and it seems to be practically undocumented, so I would guess that nobody would use that.
Other atoms that can be used (although they stem from the `ip_address` type, not `hostname`) are `loopback` and `any`, which I would treat like an IP adress. We don't have many options here, anyway, except maybe defaulting to `"localhost"` in case of `loopback`...

zuiderkwast commented:
FYI: I created the wiki page https://github.com/mysql-otp/mysql-otp/wiki/TLS which we can update once this is submitted and we have tagged a new version. We can also consider changing or removing the default `{versions, [tlsv1]}`.

juhlig commented:
> Mentioning this in EDoc for start_link/1 would be nice.

Ok, I added a note about `verify` defaulting to `verify_peer` and `server_name_indication` defaulting to the `host` option if it is a hostname string. You ok with the phrasing?

I also added a more elaborate commit message.

zuiderkwast commented:
Yes, very good!

juhlig commented:
> Yes, very good!

This means ready for merge? ;)

zuiderkwast commented:
Could have been merged long ago. :) EDoc, etc. can always be added/changed later.

juhlig commented:
> EDoc, etc. can always be added/changed later.

Well, knowing how I tend to forget those things if I don't do them in the same PR... later --> never =^^=
